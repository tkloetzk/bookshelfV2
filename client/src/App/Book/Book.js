import React from 'react'
import Card from '@material-ui/core/Card'
import CardHeader from '@material-ui/core/CardHeader'
import IconButton from '@material-ui/core/IconButton'
import CardMedia from '@material-ui/core/CardMedia'
import Typography from '@material-ui/core/Typography'
import OwnedBook from '@material-ui/icons/Home'
import UnownedBook from '@material-ui/icons/HomeOutlined'
import { makeStyles } from '@material-ui/core/styles'
import CardContent from '@material-ui/core/CardContent'
import PropTypes from 'prop-types'
import CardActions from '@material-ui/core/CardActions'
import ExpandMoreIcon from '@material-ui/icons/ExpandMore'
import Collapse from '@material-ui/core/Collapse'

const useStyles = makeStyles(theme => ({
  card: {
    width: 235,
    maxHeight: 459,
    display: 'flex',
    flexDirection: 'column',
    alignItems: 'center',
    textAlign: 'center',
    margin: '13px',
  },
  header: {
    backgroundColor: theme.palette.primary.gray,
    paddingBottom: '13px',
  },
  media: {
    width: '33%',
    paddingTop: '16px',
  },
  iconButton: {
    alignSelf: 'center',
  },
  cardHeader: {
    padding: '13px 13px 0px 13px',
  },
  headerContent: {
    paddingLeft: 9,
  },
  avatar: {
    marginRight: 0,
  },
  description: {
    height: '153px',
    overflow: 'auto',
  },
  expand: {
    transform: 'rotate(0deg)',
  },
  expandAction: {
    padding: 0,
  },
  expanded: {
    maxHeight: 574,
  },
  expandOpen: {
    transform: 'rotate(180deg)',
  },
}))

export default function Book({ book }) {
  const classes = useStyles()
  const [expanded, setExpanded] = React.useState(false)

  function handleExpandClick() {
    setExpanded(!expanded)
  }

  return (
    <Card
      className={[classes.card, expanded ? classes.expanded : null].join(' ')}
    >
      <div className={classes.header}>
        <CardHeader
          avatar={
            <Typography variant="body2">
              {(Math.round(book.adjustedRating * 1000) / 1000).toString()}
            </Typography>
          }
          action={
            <IconButton aria-label={book.owned ? 'owned' : 'unowned'}>
              {book.owned ? <OwnedBook /> : <UnownedBook />}
            </IconButton>
          }
          disableTypography
          title={
            <Typography variant="body2" align="center">
              {book.title}
            </Typography>
          }
          classes={{
            action: classes.iconButton,
            root: classes.cardHeader,
            avatar: classes.avatar,
            content: classes.headerContent,
          }}
        />
        <Typography variant="caption" align="center">
          {book.categories.join(', ')}
        </Typography>
      </div>
      {book.thumbnail && (
        <CardMedia
          classes={{ media: classes.media }}
          component="img"
          image={book.thumbnail}
          title={book.title}
        />
      )}
      <CardContent>
        <Typography
          variant="caption"
          color="textSecondary"
          component="p"
          className={classes.description}
        >
          {book.description}
        </Typography>
      </CardContent>
      <CardActions classes={{ root: classes.expandAction }}>
        <IconButton
          data-testid="expandButton"
          onClick={handleExpandClick}
          className={expanded ? classes.expandOpen : classes.expand}
        >
          <ExpandMoreIcon />
        </IconButton>
      </CardActions>
      <Collapse in={expanded} timeout="auto" unmountOnExit>
        <CardContent>
          <Typography variant="caption" component="div">
            Amazon Rating: {book.amazonAverageRating}
          </Typography>
          <Typography variant="caption" component="div">
            Goodreads Rating: {book.goodreadsAverageRating}
          </Typography>
          <Typography variant="caption" component="div">
            Amazon Review: {book.amazonRatingsCount}
          </Typography>
          <Typography variant="caption" component="div">
            Goodreads Review: {book.goodreadsRatingsCount}
          </Typography>
        </CardContent>
      </Collapse>
    </Card>
  )
}

Book.propTypes = {
  book: PropTypes.shape({
    adjustedRating: PropTypes.number,
    owned: PropTypes.bool,
    title: PropTypes.string,
    categories: PropTypes.arrayOf(PropTypes.string),
    thumbnail: PropTypes.string,
    description: PropTypes.string,
    amazonAverageRating: PropTypes.number,
    amazonRatingsCount: PropTypes.number,
    goodreadsAverageRating: PropTypes.number,
    goodreadsRatingsCount: PropTypes.number,
  }).isRequired,
}
