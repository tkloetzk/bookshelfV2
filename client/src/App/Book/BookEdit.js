import React from 'react'
import Card from '@material-ui/core/Card'
import Grid from '@material-ui/core/Grid'
import TextField from '@material-ui/core/TextField'
import IconButton from '@material-ui/core/IconButton'
import SaveIcon from '@material-ui/icons/Save'
import compareDifferences from '../../util/compareDifferences'
import Autocomplete from '@material-ui/lab/Autocomplete'
import { useSelector } from 'react-redux'
import { makeStyles } from '@material-ui/core/styles'

const useStyles = makeStyles(theme => ({
  card: {
    width: 235,
    maxHeight: 459,
    display: 'flex',
    flexDirection: 'column',
    margin: 13,
  },
}))

export default function Book({ book, setEditMode, handleSave }) {
  const classes = useStyles()
  const [bookEdit, setBookEdit] = React.useState(book)
  const genres = useSelector(state => state.bookshelf.genres)

  function setValue(e) {
    console.log(e.target)
    const newBookEdit = Object.assign({}, bookEdit, {
      [e.target.id]: e.target.value,
    })
    setBookEdit(newBookEdit)
  }

  function handleEditSave() {
    const diff = compareDifferences(book, bookEdit, [], true)
    console.log(diff)
    handleSave(bookEdit, diff)
    setEditMode(false)
  }

  function handleCancel() {
    setBookEdit(book)
    setEditMode(false)
  }

  return (
    <Card className={classes.card}>
      <Grid container justify="center">
        <Grid item xs={10}>
          <TextField
            value={bookEdit.title}
            fullWidth
            id="title"
            label="Title"
            margin="dense"
            onChange={setValue}
            inputProps={{ style: { fontSize: 12 } }}
          />
        </Grid>
        <Grid item xs={10}>
          <Autocomplete
            multiple
            options={genres.map(genre => genre.value)}
            defaultValue={bookEdit.categories}
            freeSolo
            id="categories"
            filterSelectedOptions
            disableCloseOnSelect
            onChange={(e, value) =>
              setValue({ target: { id: 'categories', value } })
            }
            renderInput={params => (
              <TextField
                {...params}
                label="Categories"
                margin="dense"
                fullWidth
              />
            )}
          />
        </Grid>
        <Grid item xs={10}>
          <TextField
            multiline
            value={bookEdit.description}
            fullWidth
            rowsMax={4}
            id="description"
            label="Description"
            margin="dense"
            onChange={setValue}
            inputProps={{ style: { fontSize: 12 } }}
          />
        </Grid>
        <Grid item xs={5}>
          <TextField
            value={bookEdit.amazonAverageRating}
            fullWidth
            type="number"
            id="amazonAverageRating"
            label="AZ Rating"
            margin="dense"
            onChange={setValue}
            inputProps={{ style: { fontSize: 12 }, step: 0.1 }}
          />
        </Grid>
        <Grid item xs={5}>
          <TextField
            value={bookEdit.goodreadsAverageRating}
            fullWidth
            type="number"
            id="goodreadsAverageRating"
            label="GR Rating"
            margin="dense"
            onChange={setValue}
            inputProps={{ style: { fontSize: 12 }, step: 0.1 }}
          />
        </Grid>
        <Grid item xs={5}>
          <TextField
            value={bookEdit.amazonRatingsCount}
            fullWidth
            type="number"
            id="amazonRatingsCount"
            label="AZ Reviews"
            margin="dense"
            onChange={setValue}
            inputProps={{ style: { fontSize: 12 } }}
          />
        </Grid>
        <Grid item xs={5}>
          <TextField
            value={bookEdit.goodreadsRatingsCount}
            fullWidth
            type="number"
            id="goodreadsRatingsCount"
            label="GR Reviews"
            margin="dense"
            onChange={setValue}
            inputProps={{ style: { fontSize: 12 } }}
          />
        </Grid>
        <Grid container>
          <Grid item xs={6}>
            <IconButton data-testid="saveButton" onClick={handleEditSave}>
              <SaveIcon />
            </IconButton>
          </Grid>
          <Grid item xs={6}>
            <IconButton data-testid="cancelButton" onClick={handleCancel}>
              <SaveIcon />
            </IconButton>
          </Grid>
        </Grid>
      </Grid>
    </Card>
  )
}
