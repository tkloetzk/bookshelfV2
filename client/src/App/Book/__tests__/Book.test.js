import { MuiThemeProvider } from '@material-ui/core/styles'
import React from 'react'
import { fireEvent, render, wait, cleanup } from '@testing-library/react'
import muiTheme from '../../../config/themeConfig'
import Book from '../Book'
import '@testing-library/jest-dom/extend-expect'

describe('Book', () => {
  let book
  beforeEach(() => {
    book = {
      categories: ['Toddler', 'Parenting'],
      read: false,
      owned: true,
      _id: '5c7893eabc9b222a3547548f',
      amazonAverageRating: 4.2,
      amazonRatingsCount: 611,
      price: '',
      isbn: '9780553381436',
      title: 'The Happiest Toddler on the Block',
      subtitle:
        'The New Way to Stop the Daily Battle of Wills and Raise a Secure and Well-Behaved One- to Four-Year-Old',
      description:
        "The pediatrician-author of The Happiest Baby on the Block offers parents a groundbreaking new approach for dealing with toddlers, drawing a parallel between a child's development and humankind's journey to civilization and presenting specific behavioral techniques designed to enhance parent-child communication, alleviate tantrums, and increase positive relationships. Reprint. 60,000 first printing.",
      thumbnail:
        'http://books.google.com/books/content?id=njQpY18IazsC&printsec=frontcover&img=1&zoom=1&source=gbs_api',
      goodreadsAverageRating: 3.5,
      goodreadsRatingsCount: 5451,
      __v: 0,
    }
  })
  afterEach(cleanup)
  describe('render', () => {
    it('should render as expected', () => {
      const { asFragment } = render(
        <MuiThemeProvider theme={muiTheme}>
          <Book book={book} />
        </MuiThemeProvider>
      )
      expect(asFragment()).toMatchSnapshot()
    })
    it('should render as expected if there is no thumbnail', () => {
      book = { ...book, thumbnail: null }
      const { asFragment } = render(
        <MuiThemeProvider theme={muiTheme}>
          <Book book={book} />
        </MuiThemeProvider>
      )
      expect(asFragment()).toMatchSnapshot()
    })
    it('should render as expected if there are differences', () => {
      book = {
        ...book,
        differences: [
          { key: 'title', currentValue: 'old title', newValue: book.title },
          {
            key: 'amazonRatingsCount',
            currentValue: 123,
            newValue: book.amazonRatingsCount,
          },
        ],
      }
      const { asFragment } = render(
        <MuiThemeProvider theme={muiTheme}>
          <Book book={book} />
        </MuiThemeProvider>
      )
      expect(asFragment()).toMatchSnapshot()
    })
  })
  describe('expand', () => {
    it('expands on handleExpandClick', async () => {
      const { asFragment, getByTestId } = render(
        <MuiThemeProvider theme={muiTheme}>
          <Book book={book} />
        </MuiThemeProvider>
      )
      await wait(() => {
        fireEvent.click(getByTestId('expandButton'))
      })

      expect(asFragment()).toMatchSnapshot()
    })
  })
  describe('IconButtons', () => {
    describe('ownedIcon', () => {
      it('shows the unowned icon for an unowned book', () => {
        book = { ...book, owned: false }
        const { asFragment } = render(
          <MuiThemeProvider theme={muiTheme}>
            <Book book={book} />
          </MuiThemeProvider>
        )

        expect(asFragment()).toMatchSnapshot()
      })
      it('calls handleSave with correct parameters', async () => {
        book = { ...book, owned: false }
        const handleSave = jest.fn()
        const { getByTestId } = render(
          <MuiThemeProvider theme={muiTheme}>
            <Book book={book} handleSave={handleSave} />
          </MuiThemeProvider>
        )

        await wait(() => {
          fireEvent.click(getByTestId('ownedIcon'))
        })
        expect(handleSave).toHaveBeenCalledWith(book, [
          { key: 'owned', newValue: true },
        ])
      })
    })
  })
})
