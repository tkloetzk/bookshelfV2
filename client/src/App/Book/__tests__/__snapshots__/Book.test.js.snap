// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Book IconButtons shows the unowned icon for an unowned book 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiPaper-elevation1 MuiCard-root makeStyles-card-233  MuiPaper-rounded"
  >
    <div
      class="MuiCardHeader-root makeStyles-cardHeader-236"
    >
      <div
        class="MuiCardHeader-avatar makeStyles-avatar-238"
      >
        <p
          class="MuiTypography-root MuiTypography-body2"
        >
          NaN
        </p>
      </div>
      <div
        class="MuiCardHeader-content makeStyles-headerContent-237"
      >
        <p
          class="MuiTypography-root MuiTypography-body2 MuiTypography-alignCenter"
        >
          The Happiest Toddler on the Block
        </p>
      </div>
      <div
        class="MuiCardHeader-action makeStyles-iconButton-235"
      >
        <button
          aria-label="unowned"
          class="MuiButtonBase-root MuiIconButton-root"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M12 5.69l5 4.5V18h-2v-6H9v6H7v-7.81l5-4.5M12 3L2 12h3v8h6v-6h2v6h6v-8h3L12 3z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
    <span
      class="MuiTypography-root MuiTypography-caption MuiTypography-alignCenter"
    >
      Toddler, Parenting
    </span>
    <img
      class="MuiCardMedia-root MuiCardMedia-media makeStyles-media-234"
      src="http://books.google.com/books/content?id=njQpY18IazsC&printsec=frontcover&img=1&zoom=1&source=gbs_api"
      title="The Happiest Toddler on the Block"
    />
    <div
      class="MuiCardContent-root"
    >
      <p
        class="MuiTypography-root makeStyles-description-239 MuiTypography-caption MuiTypography-colorTextSecondary"
      >
        The pediatrician-author of The Happiest Baby on the Block offers parents a groundbreaking new approach for dealing with toddlers, drawing a parallel between a child's development and humankind's journey to civilization and presenting specific behavioral techniques designed to enhance parent-child communication, alleviate tantrums, and increase positive relationships. Reprint. 60,000 first printing.
      </p>
    </div>
    <div
      class="MuiCardActions-root makeStyles-expandAction-241 MuiCardActions-spacing"
    >
      <button
        class="MuiButtonBase-root MuiIconButton-root makeStyles-expand-240"
        data-testid="expandButton"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Book expand expands on handleExpandClick 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiPaper-elevation1 MuiCard-root makeStyles-card-117 makeStyles-expanded-126 MuiPaper-rounded"
  >
    <div
      class="MuiCardHeader-root makeStyles-cardHeader-120"
    >
      <div
        class="MuiCardHeader-avatar makeStyles-avatar-122"
      >
        <p
          class="MuiTypography-root MuiTypography-body2"
        >
          NaN
        </p>
      </div>
      <div
        class="MuiCardHeader-content makeStyles-headerContent-121"
      >
        <p
          class="MuiTypography-root MuiTypography-body2 MuiTypography-alignCenter"
        >
          The Happiest Toddler on the Block
        </p>
      </div>
      <div
        class="MuiCardHeader-action makeStyles-iconButton-119"
      >
        <button
          aria-label="owned"
          class="MuiButtonBase-root MuiIconButton-root"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M10 20v-6h4v6h5v-8h3L12 3 2 12h3v8z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
    <span
      class="MuiTypography-root MuiTypography-caption MuiTypography-alignCenter"
    >
      Toddler, Parenting
    </span>
    <img
      class="MuiCardMedia-root MuiCardMedia-media makeStyles-media-118"
      src="http://books.google.com/books/content?id=njQpY18IazsC&printsec=frontcover&img=1&zoom=1&source=gbs_api"
      title="The Happiest Toddler on the Block"
    />
    <div
      class="MuiCardContent-root"
    >
      <p
        class="MuiTypography-root makeStyles-description-123 MuiTypography-caption MuiTypography-colorTextSecondary"
      >
        The pediatrician-author of The Happiest Baby on the Block offers parents a groundbreaking new approach for dealing with toddlers, drawing a parallel between a child's development and humankind's journey to civilization and presenting specific behavioral techniques designed to enhance parent-child communication, alleviate tantrums, and increase positive relationships. Reprint. 60,000 first printing.
      </p>
    </div>
    <div
      class="MuiCardActions-root makeStyles-expandAction-125 MuiCardActions-spacing"
    >
      <button
        class="MuiButtonBase-root MuiIconButton-root makeStyles-expandOpen-127"
        data-testid="expandButton"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
    <div
      class="MuiCollapse-container"
      style="min-height: 0px; height: 0px; transition-duration: 0ms;"
    >
      <div
        class="MuiCollapse-wrapper"
      >
        <div
          class="MuiCollapse-wrapperInner"
        >
          <div
            class="MuiCardContent-root"
          >
            <div
              class="MuiTypography-root MuiTypography-caption"
            >
              Amazon Rating: 4.2
            </div>
            <div
              class="MuiTypography-root MuiTypography-caption"
            >
              Goodreads Rating: 3.5
            </div>
            <div
              class="MuiTypography-root MuiTypography-caption"
            >
              Amazon Review: 611
            </div>
            <div
              class="MuiTypography-root MuiTypography-caption"
            >
              Goodreads Review: 5451
            </div>
          </div>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`Book render should render as expected with book prop 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiPaper-elevation1 MuiCard-root makeStyles-card-1  MuiPaper-rounded"
  >
    <div
      class="MuiCardHeader-root makeStyles-cardHeader-4"
    >
      <div
        class="MuiCardHeader-avatar makeStyles-avatar-6"
      >
        <p
          class="MuiTypography-root MuiTypography-body2"
        >
          NaN
        </p>
      </div>
      <div
        class="MuiCardHeader-content makeStyles-headerContent-5"
      >
        <p
          class="MuiTypography-root MuiTypography-body2 MuiTypography-alignCenter"
        >
          The Happiest Toddler on the Block
        </p>
      </div>
      <div
        class="MuiCardHeader-action makeStyles-iconButton-3"
      >
        <button
          aria-label="owned"
          class="MuiButtonBase-root MuiIconButton-root"
          tabindex="0"
          type="button"
        >
          <span
            class="MuiIconButton-label"
          >
            <svg
              aria-hidden="true"
              class="MuiSvgIcon-root"
              focusable="false"
              role="presentation"
              viewBox="0 0 24 24"
            >
              <path
                d="M10 20v-6h4v6h5v-8h3L12 3 2 12h3v8z"
              />
            </svg>
          </span>
          <span
            class="MuiTouchRipple-root"
          />
        </button>
      </div>
    </div>
    <span
      class="MuiTypography-root MuiTypography-caption MuiTypography-alignCenter"
    >
      Toddler, Parenting
    </span>
    <img
      class="MuiCardMedia-root MuiCardMedia-media makeStyles-media-2"
      src="http://books.google.com/books/content?id=njQpY18IazsC&printsec=frontcover&img=1&zoom=1&source=gbs_api"
      title="The Happiest Toddler on the Block"
    />
    <div
      class="MuiCardContent-root"
    >
      <p
        class="MuiTypography-root makeStyles-description-7 MuiTypography-caption MuiTypography-colorTextSecondary"
      >
        The pediatrician-author of The Happiest Baby on the Block offers parents a groundbreaking new approach for dealing with toddlers, drawing a parallel between a child's development and humankind's journey to civilization and presenting specific behavioral techniques designed to enhance parent-child communication, alleviate tantrums, and increase positive relationships. Reprint. 60,000 first printing.
      </p>
    </div>
    <div
      class="MuiCardActions-root makeStyles-expandAction-9 MuiCardActions-spacing"
    >
      <button
        class="MuiButtonBase-root MuiIconButton-root makeStyles-expand-8"
        data-testid="expandButton"
        tabindex="0"
        type="button"
      >
        <span
          class="MuiIconButton-label"
        >
          <svg
            aria-hidden="true"
            class="MuiSvgIcon-root"
            focusable="false"
            role="presentation"
            viewBox="0 0 24 24"
          >
            <path
              d="M16.59 8.59L12 13.17 7.41 8.59 6 10l6 6 6-6z"
            />
          </svg>
        </span>
        <span
          class="MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</DocumentFragment>
`;
